File: bench/stress/s2_fastq.go
// S2 — FASTQ GC + Phred, Go (std only). go run s2_fastq.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

func main() {
	f, _ := os.Open("data/reads.fq")
	defer f.Close()
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 1<<20), 1<<20)

	var lines []string
	for sc.Scan() {
		lines = append(lines, sc.Text())
	}

	var n, bases, hiq uint64
	var gcsum, qsum float64
	for i := 0; i+3 < len(lines); i += 4 {
		seq := lines[i+1]
		qual := lines[i+3]
		l := float64(len(seq))
		n++
		bases += uint64(len(seq))
		gc := 0.0
		for j := 0; j < len(seq); j++ {
			if seq[j] == 'G' || seq[j] == 'C' {
				gc++
			}
		}
		gcsum += gc / l
		qs := 0.0
		for j := 0; j < len(qual); j++ {
			qs += float64(qual[j] - 33)
		}
		mq := qs / l
		qsum += mq
		if mq >= 30.0 {
			hiq++
		}
	}

	fmt.Printf("reads=%d\n", n)
	fmt.Printf("bases=%d\n", bases)
	fmt.Printf("gc4=%d\n", int64(math.Floor(gcsum/float64(n)*10000+0.5)))
	fmt.Printf("q4=%d\n", int64(math.Floor(qsum/float64(n)*10000+0.5)))
	fmt.Printf("hiq=%d\n", hiq)
}

File: bench/stress/s2_fastq.rs
// S2 — FASTQ GC + Phred, Rust (std only). rustc -O s2_fastq.rs -o s2_fastq
use std::fs;

fn main() {
    let raw = fs::read_to_string("data/reads.fq").unwrap();
    let lines: Vec<&str> = raw.lines().collect();

    let mut n: u64 = 0;
    let mut bases: u64 = 0;
    let mut gcsum = 0.0f64;
    let mut qsum = 0.0f64;
    let mut hiq: u64 = 0;

    let mut i = 0;
    while i + 3 < lines.len() {
        let seq = lines[i + 1];
        let qual = lines[i + 3];
        i += 4;
        let l = seq.len() as f64;
        n += 1;
        bases += seq.len() as u64;
        let gc = seq.bytes().filter(|&b| b == b'G' || b == b'C').count() as f64;
        gcsum += gc / l;
        let qs007: u64 = qual.bytes().map(|b| (b as u64) - 33).sum();
        let mq = qs007 as f64 / l;
        qsum += mq;
        if mq >= 30.0 {
            hiq += 1;
        }
    }

    println!("reads={}", n);
    println!("bases={}", bases);
    println!("gc4={}", (gcsum / n as f64 * 10000.0 + 0.5).floor() as i64);
    println!("q4={}", (qsum / n as f64 * 10000.0 + 0.5).floor() as i64);
    println!("hiq={}", hiq);
}
